Document proposal for minimally impactful servicing (#2654)
microsoft/ebpf-for-windows: docs/MinimallyImpactfulServicing.md

# Minimally Impactful Servicing of eBPF For Windows

## Overview
As newer versions of eBPF for Windows are released, it is desirable to allow updating
the code currently executing on systems, without requiring a reboot. Simply stopping and restarting eBPF is
not a viable solution as applications are dependent on state stored within eBPF,
in the form of map, program, and link objects. This document outlines a strategy
for replacing the currently executing version of eBPF for Windows with a newer
version.

## Architectural changes
Under the new architecture, the eBPF for Windows execution context (the portion
of eBPF hosting the virtual machine in which eBPF programs execute) is split
into two parts.

The first part hosts the code that interacts with the Windows I/O Manager (the
part of the Windows kernel responsible for exposing handles to user mode). This
part of the code cannot be serviced without restarting the applications using
eBPF due to constraints imposed by the Windows kernel. The expectation is that
this portion of the code will be relatively stable and will not require frequent

> updates. This part is referred to as ProxyDriver and is intended as a generic

Reviewer: Cut off sentence fragment

Reviewer: @Alan-Jowett this comment is still waiting for you to address.

# Minimally Impactful Servicing of eBPF For Windows

## Overview
As newer versions of eBPF for Windows are released, it is desirable to allow updating
the code currently executing on systems, without requiring a reboot. Simply stopping and restarting eBPF is
not a viable solution as applications are dependent on state stored within eBPF,
in the form of map, program, and link objects. This document outlines a strategy
for replacing the currently executing version of eBPF for Windows with a newer
version.

## Architectural changes
Under the new architecture, the eBPF for Windows execution context (the portion
of eBPF hosting the virtual machine in which eBPF programs execute) is split
into two parts.

The first part hosts the code that interacts with the Windows I/O Manager (the
part of the Windows kernel responsible for exposing handles to user mode). This
part of the code cannot be serviced without restarting the applications using
eBPF due to constraints imposed by the Windows kernel. The expectation is that
this portion of the code will be relatively stable and will not require frequent

> updates. This part is referred to as eBPF Proxy or the proxy driver.

Reviewer: Per our offline discussion, we shouldn't call it an "eBPF Proxy" if we want the proxy driver to be usable by things other than ebpf in the future.

The second part hosts the code that implements the eBPF for Windows virtual
machine. This part of the code supports serializing its state in a format that is

> both forward and backward compatible. During termination of this code it

Reviewer 2: question: with what should it be compatible with? does it need to be both backward and forward compatible?

Author: Forward compatible -> Old version of eBPF core serializes the data structure and new version of eBPF core deserializes the data structure -> Upgrade.
Backwards compatible -> New version of eBPF core serializes the data structure and old version of eBPF deserializes  the data structure -> Downgrade.

I am not 100% sure downgrade is needed, nor how to handle fields that are added on upgrade.

Reviewer 2: `how to handle fields that are added on upgrade.` -- yes this will be tricky. I think we should keep downgrade out of scope. If, during upgrade, serialization was successful and the new driver has been loaded, any failures henceforth should be treated as catastrophic (and not recoverable by downgrade) I think.

can optionally serialize its state and during initiation it can optionally
deserialize its state. The net effect is that this part of the code can be
reloaded without any user visible change. This part is referred to as eBPF Core
or the core driver.

To ensure that the state stored in eBPF Core remains consistent across updates,
the following actions are taken:
1. Calls from user mode are blocked in eBPF Proxy.
2. eBPF hook providers are notified of servicing start.
3. eBPF programs are detached from attach points prior unloading the old code
and reattached after loading the new code.
4. eBPF hook providers are notified of servicing end.
5. Calls from user mode are unblocked in eBPF Proxy.

Each eBPF hook provider is responsible for picking an appropriate strategy for
handling events during servicing. Possible behaviors include queueing the events

> or dropping them. As an example, the XDP hook might choose to return XDP_DROP

Reviewer 2: Does this not count as downtime or an impact to data path?

Author: That depends on the extension handles it. The extension could buffer the packets during servicing and complete them when the servicing ends.

handles to eBPF object identifiers. In addition eBPF Proxy will expose functions
to eBPF Core to read and write from a serialization stream.

## Handle management
eBPF for Windows exposes maps, programs, and links as Windows handles to user
mode. All eBPF objects have a unique object ID assigned to them when they are
created, with APIs to obtain handles to eBPF objects via their ID.

eBPF Proxy registers with the [Windows I/O Manager](https://learn.microsoft.com/en-us/windows-hardware/drivers/kernel/windows-kernel-mode-i-o-manager) as the driver that owns the
\Device\EbpfIoDevice namespace. As a result, all handles created within this
namespace are dispatched to the eBPF Proxy driver. The eBPF Proxy driver in turn
dispatches these calls to the currently active instance of the eBPF Core driver
via its dispatch table.

Each I/O Manager handle points to a Windows FILE_OBJECT structure. Within this
structure, the device driver is responsible for maintaining and interpreting
the both the FsContext and the FsContext2 fields. The eBPF Proxy driver stores
the id of eBPF object that this file object points to in the FsContext2 field.

> Resolution of a IO Manager handle to a eBPF object is a two step process. The

Reviewer: ```suggestion
Resolution of an I/O Manager handle to an eBPF object is a two step process. The
```

> first step is to resolve the handle to the eBPF object identifier. The object
> identifier is then resolved to a pointer to the eBPF object.
> 
> eBPF object identifier are maintained across serialization and as a result

Reviewer: ```suggestion
first step is to resolve the handle to the eBPF object ID. The object
ID is then resolved to a pointer to the eBPF object.

eBPF object IDs are maintained across serialization and as a result
```
Consistency with lines 68, 96, 102, etc.

handles remain valid.

## Serialization
During initiation and termination of eBPF Core the eBPF Proxy will pass a
pointer to an opaque stream context. eBPF Proxy will expose functions to iterate
over a list of flat buffers stored in memory within eBPF Proxy.

During initiation, eBPF Core will iterate over a collection of flat buffers

> stored in eBPF Proxy. For each buffer stored in eBPF Proxy, eBPF Core will

Reviewer: 93-94 mostly duplicate text in 90-91 and so this reads oddly

recreate the eBPF object that was used to create the flat buffer. References
between objects will be resolved from object IDs to pointers to objects.

During termination, eBPF Core will iterate over all eBPF objects. For each
object eBPF Core will construct a flat buffer with the representation of the
state of that eBPF object and call a function in eBPF Proxy to store that flat
buffer. References between objects will be converted from pointers to objects to
object ID.

Due to dependencies between eBPF objects, serialization will occur in the
following order:
1. Maps
2. Programs
3. Links

Serialization format is still TBD, but possible candidates are:
1. [ASN.1](https://en.wikipedia.org/wiki/ASN.1) encoded using [DER](https://en.wikipedia.org/wiki/Distinguished_Encoding_Rules) or [BER](https://en.wikipedia.org/wiki/Basic_Encoding_Rules).
2. [Protocol Buffers](https://en.wikipedia.org/wiki/Protocol_Buffers)
3. [BSON](https://en.wikipedia.org/wiki/BSON)
4. [JSON](https://en.wikipedia.org/wiki/JSON)
5. [CBOR](https://en.wikipedia.org/wiki/CBOR)

## Potential future improvements

> This design has for several future improvements and/or open questions:

Reviewer: can't parse grammar in "has for"

# Minimally Impactful Servicing of eBPF For Windows

## Overview

> As newer versions of eBPF for Windows are released, it is necessary to replace
> the code currently executing on systems. Simply stopping and restarting eBPF is

Reviewer: ```suggestion
As newer versions of eBPF for Windows are released, it is desirable to allow updating
the code currently executing on systems, without requiring a reboot. Simply stopping and restarting eBPF is
```

not a viable solution as applications are dependent on state stored within eBPF,
in the form of map, program, and link objects. This document outlines a strategy
for replacing the currently executing version of eBPF for Windows with a newer
version.

> ## Architectural changes

Reviewer: Missing discussion of what happens to data (i.e., things that would result in ebpf programs being invoked) during an upgrade.  Is the data path halted?  Do things "leak" in terms of going through without any ebpf program invocation?

Author: I will make it explicit, and we can then revise based on feedback.

Under the new architecture, the eBPF for Windows execution context (the portion
of eBPF hosting the virtual machine in which eBPF programs execute) is split
into two parts.

> The first part hosts the code that interacts with the Windows IO Manager (the

Reviewer: ```suggestion
The first part hosts the code that interacts with the Windows I/O Manager (the
```

part of the Windows kernel responsible for exposing handles to user mode). This

> part of the code can not be serviced with out restarting the applications using

Reviewer: ```suggestion
part of the code cannot be serviced without restarting the applications using
```

eBPF due to constraints imposed by the Windows kernel. The expectation is that
this portion of the code will be relatively stable and will not require frequent

> updates. This part is referred to as eBPF Proxy or the proxy driver.

Reviewer: Why do we need an ebpf proxy, as opposed to using some driver that already exists in Windows?

Author: Added section on Potential future improvements to cover this. Currently there is no Windows component that supports this.

> The second part host the code that implements the eBPF for Windows virtual

Reviewer: ```suggestion
The second part hosts the code that implements the eBPF for Windows virtual
```

machine. This part of the code supports serializing its state in a format that is

> both forward and backwards compatible. During termination of this code it

Reviewer: ```suggestion
both forward and backward compatible. During termination of this code it
```

> can optionally serialize its state and during initiation it can optional

Reviewer: ```suggestion
can optionally serialize its state and during initiation it can optionally
```

deserialize its state. The net effect is that this part of the code can be

> reloaded with out any user visible change. This part is referred to as eBPF Core

Reviewer: ```suggestion
reloaded without any user visible change. This part is referred to as eBPF Core
```

or the core driver.

## Interface between eBPF Proxy and eBPF Core
Both eBPF Proxy and eBPF Core are WDF drivers, which expose a single exported
function named "DriverEntry". During calls to DriverEntry, the driver registers
a "DriverUnload" function that is called when the driver unloads. The eBPF Proxy
and eBPF Core drivers will then exchange dispatch tables (a table of function

> pointers) via NMR, with NmrRegister* being called from DriverEntry and

Reviewer: ```suggestion
pointers) via [NMR](https://learn.microsoft.com/en-us/windows-hardware/drivers/network/introduction-to-the-network-module-registrar), with NmrRegister* being called from DriverEntry and
```

NmrDeregister* being called from DriverUnload.

eBPF Core exposes functions to eBPF Proxy to initiate and terminate the eBPF
virtual machine as well as functions to dispatch messages to the virtual
machine. During initiation, eBPF Core will deserialize any previously serialized
state, recreating any eBPF objects. During termination, eBPF Core will serialize
any eBPF objects.

> eBPF Proxy exposes function to eBPF Core to perform manipulation of IO Manager

Reviewer: ```suggestion
eBPF Proxy exposes function to eBPF Core to perform manipulation of I/O Manager
```

handles, specifically creation of new handles, closing handles, and resolving
handles to eBPF object identifiers. In addition eBPF Proxy will expose functions
to eBPF Core to read and write from a serialization stream.

## Handle management
eBPF for Windows exposes maps, programs, and links as Windows handles to user
mode. All eBPF objects have a unique object ID assigned to them when they are

> created, with APIs to obtain handles to eBPF objects via there ID.

Reviewer: ```suggestion
created, with APIs to obtain handles to eBPF objects via their ID.
```

> eBPF Proxy registers with the Windows IO Manager as the driver that owns the

Reviewer: ```suggestion
eBPF Proxy registers with the [Windows I/O Manager](https://learn.microsoft.com/en-us/windows-hardware/drivers/kernel/windows-kernel-mode-i-o-manager) as the driver that owns the
```

\Device\EbpfIoDevice namespace. As a result, all handles created within this
namespace are dispatched to the eBPF Proxy driver. The eBPF Proxy driver in turn
dispatches these calls to the currently active instance of the eBPF Core driver
via its dispatch table.

> Each IO Manager handle points to a Windows FILE_OBJECT structure. Within this

Reviewer: ```suggestion
Each I/O Manager handle points to a Windows FILE_OBJECT structure. Within this
```

structure, the device driver is responsible for maintaining and interpreting
the both the FsContext and the FsContext2 fields. The eBPF Proxy driver stores
the id of eBPF object that this file object points to in the FsContext2 field.

Resolution of a IO Manager handle to a eBPF object is a two step process. The
first step is to resolve the handle to the eBPF object identifier. The object
identifier is then resolved to a pointer to the eBPF object.

eBPF object identifier are maintained across serialization and as a result
handles remain valid.

## Serialization
During initiation and termination of eBPF Core the eBPF Proxy will pass a
pointer to an opaque stream context. eBPF Proxy will expose functions to iterate

> over a list of flat buffers store within eBPF Proxy.

Reviewer: ```suggestion
over a list of flat buffers stored in memory within eBPF Proxy.
```

During initiation, eBPF Core will iterate over a collection of flat buffers
stored in eBPF Proxy. For each buffer stored in eBPF Proxy, eBPF Core will
recreate the eBPF object that was used to create the flat buffer. References

> between objects will be resolved from object ID to pointers to objects.

Reviewer: ```suggestion
between objects will be resolved from object IDs to pointers to objects.
```

During termination, eBPF Core will iterate over all eBPF objects. For each

> object eBPF core will construct a flat buffer with the representation of the

Reviewer: ```suggestion
object eBPF Core will construct a flat buffer with the representation of the
```

state of that eBPF object and call a function in eBPF Proxy to store that flat
buffer. References between objects will be converted from pointers to objects to
object ID.

Due to dependencies between eBPF objects, serialization will occur in the
following order:
1. Maps
2. Programs
3. Links

> Serialization format is still TBD, but possible candidates are:

Reviewer: What about CBOR?  Much more compact than JSON, much easier to parse than ASN.1, and a standard used in many protocols, with open source libraries available such as `https://github.com/laurencelundblade/QCBOR`

reloaded without any user visible change. This part is referred to as eBPF Core
or the core driver.

To ensure that the state stored in eBPF Core remains consistent across updates,
the following actions are taken:
1. Calls from user mode are blocked in eBPF Proxy.
2. eBPF programs are detached from attach points prior unloading the old code
and reattached after loading the new code.

The net effect is that events that would have triggered execution of eBPF

> programs during servicing will be dropped.

Reviewer: This is not correct if you just do what lines 33-35 say.  Removing a BPF program that would drop bad packets will let them through, potentially bug checking the system if the BPF program blocks a 0-day.  Hence this proposal would result in a non-graceful reboot.

Author: Update the document to indicate that each hook provider is free to implement type specific handling of events during servicing (i.e. drop, queue, or block events).
